ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/editor/plugin/ExecutionEnvironmentSection.java

```diff
@@ -324,6 +350,28 @@ private void handleAdd() {
 		}
 	}
 
+	private void handleUpgrade() {
+
+		RequiredExecutionEnvironmentHeader header = getHeader();
+		IExecutionEnvironment[] executionEnvironments = JavaRuntime.getExecutionEnvironmentsManager()
+				.getExecutionEnvironments();
+		if (executionEnvironments.length > 0) {
+			ExecutionEnvironment[] addedExecutionEnvironments = header.getEnvironments();
+			// Remove existing EE ( Upgrade is allowed only when single Java EE
+			// is present in the header)
+			getHeader().removeExecutionEnvironment(addedExecutionEnvironments[0]);
+
+			SortedMap<String, IExecutionEnvironment> executionEnvironmentMap = new TreeMap<>(new VersionComparator());
+			Arrays.asList(executionEnvironments).forEach(executionEnvironment -> executionEnvironmentMap
+					.put(executionEnvironment.getId(), executionEnvironment));
+			// add highest Java EE to the header
+			header.addExecutionEnvironment(executionEnvironmentMap.get(executionEnvironmentMap.firstKey()));
```

Comment:
The current implementation of the `Upgrade` button just upgrades to the latest version supported by PDE/JDT.
While I really like the upgrade in one step (compared to add with subsequent remove and CP update), I think we should present a dialog to select the upgrade target, similar to the dialog that is presented when one clicks add.

At the moment clicking Upgrade moves a bundle to `JavaSE-23` that will be released in September. I assume that's not what most users will target in the meantime. And in general one does not always want to upgrade to the latest but for example from one LTS like 17 to a later one like Java-21. Without the ability to select the target this new features would often not be useable.

Reply:
I agree that the user should be asked for what is to be upgraded, I remember we specifically discussed this in the issue itself, this will also remove any need to sort the list as the dialog itself is sorted already.